Support non-constant exception values in JIT

guilhermeleobas opened the pull request:
Fixes #7834

njriasan commented on a.py:
```diff
@@ -0,0 +1,52 @@
+from numba import njit
```
I assume this file will be removed once the PR is merged.

njriasan commented on numba/core/callconv.py:
```diff
@@ -396,6 +397,81 @@ def return_user_exc(self, builder, exc, exc_args=None, loc=None,
+    def set_dynamic_user_exc(self, builder, exc, exc_args, loc=None,
+                             func_name=None):
+        if not issubclass(exc, BaseException):
+            raise TypeError("exc should exception class, got %r"
```
```suggestion
            raise TypeError("exc should be an exception class, got %r"
```

njriasan commented on numba/core/callconv.py:
```diff
@@ -396,6 +397,81 @@ def return_user_exc(self, builder, exc, exc_args=None, loc=None,
+        alloc_fnty = ir.FunctionType(int8_t.as_pointer(), [int32_t])
+        alloc_fn = cgutils.get_or_insert_function(builder.module, alloc_fnty, name="malloc")
+        struct_size = int32_t(self.context.get_abi_sizeof(excinfo_t))
+        excinfo = builder.bitcast(builder.call(alloc_fn, [struct_size]), excinfo_ptr_t)
```
Is it possible to add a check if this malloc failed?

njriasan commented on numba/core/consts.py:
```diff
@@ -93,8 +93,17 @@ def _infer_call(self, func, expr):
+                new_a = None
+                try:
+                    new_a = self.infer_constant(a.name, loc=expr.loc)
+                except ConstantInferenceError as cie:
```
Is there any reason to always try infer constants when we have an exception? Is this to handle some of the existing unsupported cases?

If so I'd add a comment explaining that here and that we can still handle non-constant strings.

guilhermeleobas replied:
> Is there any reason to always try infer constants when we have an exception?

I think this is for historic reasons as Numba can only handle exceptions is constant values, until this PR lands.

njriasan commented on numba/core/ir.py:
```diff
@@ -754,6 +754,35 @@ def get_targets(self):
+    """
+    Raise an exception class and some argument *values* unknown at compile-time.
+    Note that if *exc_class* is None, a bare "raise" statement is implied
+    (i.e. re-raise the current exception).
```
What are the situations in which someone can reraise an exception? I see that code like this below won't work because we can't do `as e`
```python3
import numba

@numba.njit
def f(lst):
    try:
        idx = lst[10]
    except KeyError as e:
        raise e

f([1])
```

njriasan commented on numba/core/serialize.py:
```diff
@@ -60,6 +62,22 @@ def dumps(obj):
+                 for i, runtime_arg in enumerate(runtime_args)]
+    obj = (exc, tuple(real_args), locinfo)
+    data = dumps(obj)
+    # Does it worth computing the hash?
```
This is a good question. Adding this comment for visibility in future reviews.

njriasan reviewed (commented):
Thanks @guilhermeleobas this looks really good.

stuartarchibald reviewed (commented):
Thanks for the patch @guilhermeleobas, it's good to see this feature added. I've given this patch an initial review, there's a lot of use cases that currently fail that I think should be addressed ahead of a more thorough review of the code/implementation. Thanks again!

guilhermeleobas commented on numba/core/callconv.py:
```diff
@@ -396,6 +396,83 @@ def return_user_exc(self, builder, exc, exc_args=None, loc=None,
+
+        excinfo = builder.bitcast(ret, excinfo_ptr_t)
+
+        # bug on NRT not being enabled!?
+        # RE: No! _get_code_point(a, i) disables NRT and we cannot use nrt.allocate
+        # when NRT is disabled.
```
Sorry, it is not related. I think the error is because of https://github.com/numba/numba/issues/8236

guilhermeleobas commented on numba/core/callconv.py:
```diff
@@ -396,6 +396,83 @@ def return_user_exc(self, builder, exc, exc_args=None, loc=None,
+        # hash is computed at runtime, after all arguments are known
+        zero, one, two = int32_t(0), int32_t(1), int32_t(2)
+        builder.store(ptr, builder.gep(excinfo, [zero, zero]))
+        builder.store(builder.trunc(sz, int32_t), builder.gep(excinfo, [zero, one]))
```
I've added an overflow check

guilhermeleobas commented on numba/core/callconv.py:
```diff
@@ -396,6 +396,83 @@ def return_user_exc(self, builder, exc, exc_args=None, loc=None,
+
+        v = builder.extract_value(builder.load(struct_gv_ptr), 0)
+        _len = builder.extract_value(builder.load(struct_gv_ptr), 1)
+        pybytes = pyapi.bytes_from_string_and_size(
```
Fixed

guilhermeleobas commented on numba/tests/test_exceptions.py:
```diff
@@ -325,6 +341,43 @@ def test_user_code_error_traceback_objmode(self):
+        no_pyobj_flags.nrt = True
+        self.check_raise_instance_with_runtime_args(flags=no_pyobj_flags)
+        no_pyobj_flags.nrt = False
+
```
The cases you marked as unhashable are not related to this PR as Numba type them as static raise. As for the other four cases, the errors are correctly treated.

stuartarchibald commented on numba/core/callconv.py:
```diff
@@ -396,6 +396,83 @@ def return_user_exc(self, builder, exc, exc_args=None, loc=None,
+        alloc_fnty = ir.FunctionType(int8_t.as_pointer(), [int32_t])
+        malloc_fn = cgutils.get_or_insert_function(builder.module, alloc_fnty, name="malloc")
+        struct_size = int32_t(self.context.get_abi_sizeof(excinfo_t))
+        ret = builder.call(malloc_fn, [struct_size])
```
The NRT allocator is this:
https://github.com/numba/numba/blob/05e5e3407ba0ff3fb454943e03f6d76255556cbe/numba/core/runtime/context.py#L52-L72
which calls `NRT_Allocate` and ends up in here:
https://github.com/numba/numba/blob/05e5e3407ba0ff3fb454943e03f6d76255556cbe/numba/core/runtime/nrt.cpp#L518-L536
which is essentially just a call to `malloc`. @sklam do you have a view on what to do with this? Perhaps a meminfo could be used.

guilhermeleobas replied:
So, the memory is still leaking?

guilhermeleobas replied:
Hi @stuartarchibald, once you have the chance, can you review this PR? I believe I have addressed the memory leak issue

sklam commented:
Hm... the failing test is segfaulting at a malloc in refprune?!:

```

Program received signal SIGSEGV, Segmentation fault.
0x00007f3fe1d06bc6 in malloc () from /lib64/libc.so.6
(gdb) bt
#0  0x00007f3fe1d06bc6 in malloc () from /lib64/libc.so.6
#1  0x00007f3fdf047e4f in operator new (sz=48)
    at /home/builder/ktietz/cos6/ci_cos6/ctng-compilers_1622658800915/work/.build/x86_64-conda-linux-gnu/src/gcc/libstdc++-v3/libsupc++/new_op.cc:50
#2  0x00007f3fdc967ca5 in LLVMPY_AddRefPrunePass (PM=0x55f737c20060, subpasses=15, subgraph_limit=1000) at custom_passes.cpp:1175
#3  0x00007f3fe17b29dd in ffi_call_unix64 () from /root/miniconda3/envs/numba_dev/lib/python3.9/lib-dynload/../../libffi.so.7
#4  0x00007f3fe17b2067 in ffi_call_int () from /root/miniconda3/envs/numba_dev/lib/python3.9/lib-dynload/../../libffi.so.7
#5  0x00007f3fe17cc0f6 in _call_function_pointer (argtypecount=<optimized out>, argcount=3, resmem=0x7ffd51bd3710,
    restype=<optimized out>, atypes=<optimized out>, avalues=<optimized out>,
    pProc=0x7f3fdc967c30 <LLVMPY_AddRefPrunePass(LLVMPassManagerRef, int, size_t)>, flags=4353)
    at /usr/local/src/conda/python-3.9.5/Modules/_ctypes/callproc.c:920
```

guilhermeleobas commented on numba/core/callconv.py:
```diff
@@ -411,6 +431,231 @@ def return_user_exc(self, builder, exc, exc_args=None, loc=None,
+        objs = []
+        for i, typ in enumerate(nb_types):
+            val = builder.extract_value(builder.load(st_ptr), i)
+            self.context.nrt.decref(builder, typ, val)
```
Thanks, siu! That solved the issue!

guilhermeleobas commented:
The current flake8 failure doesn't make much sense, as I already addressed the issue it is reporting.

stuartarchibald reviewed (commented):
Thanks for the updates @guilhermeleobas, I've given the new methodology/code a review, I think it resolves some of the issues that were present in the previous iteration (need for the GIL and various leaks). I've left quite a few comments in-line, most should be relatively quick to resolve. 

The remaining things on my TODO list for this PR are:
1. Manually test this patch for usability.
2. Check the unit testing for how well it covers the proposed changes.

However, I thought I'd leave the review now as those items shouldn't impact starting to address any changes needed. Many thanks for working on this!

guilhermeleobas commented on numba/core/cpu.py:
```diff
@@ -107,7 +107,7 @@ def with_aot_codegen(self, name, **aot_options):
         return self._internal_codegen
 
-    @cached_property
+    @property
```
Yes, #8236

guilhermeleobas commented on numba/core/callconv.py:
```diff
@@ -411,6 +431,230 @@ def return_user_exc(self, builder, exc, exc_args=None, loc=None,
+            builder.store(arg, builder.gep(excinfo_p, [zero, int32_t(idx)]))
+        builder.store(excinfo_p, excinfo_pp)
+
+    def return_non_const_user_exc(self, builder, exc, exc_args, nb_types,
+                                  loc=None, func_name=None):
```
Fixed. I've renamed `non_const` -> `dynamic`

guilhermeleobas commented on numba/core/pythonapi.py:
```diff
@@ -1365,14 +1379,17 @@ def serialize_uncached(self, obj):
-        as a pointer to a {i8* data, i32 length}, structure constant
-        (suitable for passing to unserialize()).
+        as a pointer to a
+        {i8* data, i32 length, i8* hashbuf, i8* fn_ptr, i32 flag},
```
Fixed. I've renamed all occurrences to `alloc_flag`

guilhermeleobas commented on numba/core/callconv.py:
```diff
@@ -411,6 +431,230 @@ def return_user_exc(self, builder, exc, exc_args=None, loc=None,
+        #                       ^^^  ^^^  ^^^
+        #                       dynamic exc only - first 2 args are used for
+        #                                          static info
+        # Comment below details how the struct is used in the case of a dynamic
+        # exception
```
Done. I've included a comment on the method for the static function as well.

guilhermeleobas commented on numba/core/serialize.py:
```diff
@@ -60,6 +61,19 @@ def dumps(obj):
     return pickled
 
 
+def runtime_build_excinfo_struct(static_exc, exc_args):
```
Renamed one of them to `build_dynamic_excinfo_struct`

guilhermeleobas commented on numba/core/callconv.py:
```diff
@@ -411,6 +431,230 @@ def return_user_exc(self, builder, exc, exc_args=None, loc=None,
+        excinfo = api.runtime_build_excinfo_struct(static_exc_bytes, py_tuple)
+
+        # At this point, one can free the entire excinfo_ptr struct
+        if self.context.enable_nrt:
```
No, dynamic exceptions require NRT

stuartarchibald replied:
Should this then raise if the NRT is not enabled?

guilhermeleobas replied:
The NRT module already raises an exception:
```
...
  File "/Users/guilhermeleobas/git/numba/numba/core/runtime/context.py", line 46, in wrap
    memptr = func(self, builder, *args, **kwargs)
  File "/Users/guilhermeleobas/git/numba/numba/core/runtime/context.py", line 59, in allocate
    return self.allocate_unchecked(builder, size)
  File "/Users/guilhermeleobas/git/numba/numba/core/runtime/context.py", line 66, in allocate_unchecked
    self._require_nrt()
  File "/Users/guilhermeleobas/git/numba/numba/core/runtime/context.py", line 41, in _require_nrt
    raise errors.NumbaRuntimeError("NRT required but not enabled")
numba.core.errors.NumbaRuntimeError: Failed in nopython mode pipeline (step: native lowering)
NRT required but not enabled
During: lowering "<dynamic> raise <class 'ValueError'>(Var(i, a.py:7))" at /Users/guilhermeleobas/git/numba/a.py (10)
```

guilhermeleobas commented on numba/core/callconv.py:
```diff
@@ -411,6 +431,230 @@ def return_user_exc(self, builder, exc, exc_args=None, loc=None,
+            if obj == cgutils.get_null_value(obj.type):
+                # When not supported, abort compilation
+                msg = f'cannot convert native {typ} to python object'
+                raise errors.TypingError(msg)
```
No. The `TypingError` is raised at compile-time. Thus, no allocations were made at this point.

guilhermeleobas commented on numba/core/callconv.py:
```diff
@@ -411,6 +431,230 @@ def return_user_exc(self, builder, exc, exc_args=None, loc=None,
+        self.context.nrt.free(builder, fn.args[0])
+
+        # Create a tuple of CPython objects
+        tup = api.tuple_pack(objs)
```
Done in `CPUCallConv::unpack_dynamic_exception`

guilhermeleobas commented on numba/core/callconv.py:
```diff
@@ -411,6 +431,230 @@ def return_user_exc(self, builder, exc, exc_args=None, loc=None,
+
+    def unpack_exception(self, builder, api, status):
+        # Emit code that checks the alloc flag (last excinfo member)
+        # if flag > 0:
```
More than a static/dynamic switch, `flag` indicates the number of runtime arguments in an exception. For instance, if `flag == 3`, it means there are three dynamic arguments.

guilhermeleobas commented on numba/core/consts.py:
```diff
@@ -93,8 +93,18 @@ def _infer_call(self, func, expr):
+                    new_a = self.infer_constant(a.name, loc=expr.loc)
+                except ConstantInferenceError as cie:
+                    if _exc:
+                        # handle non-constant arguments
+                        new_a = a
```
You are correct. I was breaking the contract of constant inference by returning a list with possible non-constant arguments. To address it, I created a new FunctionPass called `RewriteDynamicRaises` which runs **before** type inference and translate the remaining exceptions nodes (`ir.Static`, `ir.TryStatic`) to `ir.DynamicRaises`

guilhermeleobas commented:
Hi Stuart, I believe I have addressed *almost* all your comments. There is one remaining issue I left for a near future.

I left a comment in the items I thought were important but I’ll include a list of changes below:

1. Renamed all occurrences of `non_const` to `dynamic`
2. Renamed one of the occurrences of `runtime_build_excinfo_struct` to `build_dynamic_excinfo_struct`
3. NRT is required in dynamic exceptions, so I included a test case for this scenario
4. I created a function pass called `RewriteDynamicRaises`, which converts the remaining `ir.Raise`/`ir.TryRaise` nodes into `ir.DynamicRaise`. This was necessary to avoid breaking the constant inference code.

To-Do:

1. You commented that `tup = api.tuple_pack(items)` is not checked for success. I still need to address this one.

stuartarchibald commented:
RE CI failures, looks like: `'numba.tests.test_generators.TestGenerators.test_gen4'` is causing a segfault on Python 3.8. The other is a `flake8` issue: `numba/core/callconv.py:497:81: E501 line too long (82 > 80 characters)`

stuartarchibald commented:
Commit 8baa261 had a different segfault: `'numba.tests.test_generators.TestGenerators.test_gen7_objmode'` on linux `py38_np118_cov`.

guilhermeleobas commented:
I can reproduce the failure on macos with:
```bash
$ ./runtests.py -v -f numba.tests.test_generators.TestGenerators -k objmode
test_gen1_objmode (numba.tests.test_generators.TestGenerators) ... ok
test_gen2_objmode (numba.tests.test_generators.TestGenerators) ... ok
test_gen3_objmode (numba.tests.test_generators.TestGenerators) ... ok
test_gen4_objmode (numba.tests.test_generators.TestGenerators) ... Fatal Python error: Segmentation fault

Current thread 0x0000000100f38580 (most recent call first):
  File "/Users/guilhermeleobas/git/numba/numba/tests/test_generators.py", line 205 in test_gen4_objmode
  File "/Users/guilhermeleobas/miniconda3/envs/numba/lib/python3.9/unittest/case.py", line 550 in _callTestMethod
  File "/Users/guilhermeleobas/miniconda3/envs/numba/lib/python3.9/unittest/case.py", line 592 in run
  File "/Users/guilhermeleobas/miniconda3/envs/numba/lib/python3.9/unittest/case.py", line 651 in __call__
  File "/Users/guilhermeleobas/miniconda3/envs/numba/lib/python3.9/unittest/suite.py", line 122 in run
  File "/Users/guilhermeleobas/miniconda3/envs/numba/lib/python3.9/unittest/suite.py", line 84 in __call__
  File "/Users/guilhermeleobas/miniconda3/envs/numba/lib/python3.9/unittest/runner.py", line 184 in run
  File "/Users/guilhermeleobas/git/numba/numba/testing/main.py", line 169 in run
  File "/Users/guilhermeleobas/miniconda3/envs/numba/lib/python3.9/unittest/main.py", line 271 in runTests
  File "/Users/guilhermeleobas/git/numba/numba/testing/main.py", line 361 in run_tests_real
  File "/Users/guilhermeleobas/git/numba/numba/testing/main.py", line 376 in runTests
  File "/Users/guilhermeleobas/miniconda3/envs/numba/lib/python3.9/unittest/main.py", line 101 in __init__
  File "/Users/guilhermeleobas/git/numba/numba/testing/main.py", line 204 in __init__
  File "/Users/guilhermeleobas/git/numba/numba/testing/__init__.py", line 54 in run_tests
  File "/Users/guilhermeleobas/git/numba/numba/testing/_runtests.py", line 25 in _main
  File "/Users/guilhermeleobas/git/numba/numba/runtests.py", line 9 in <module>
  File "/Users/guilhermeleobas/miniconda3/envs/numba/lib/python3.9/runpy.py", line 87 in _run_code
  File "/Users/guilhermeleobas/miniconda3/envs/numba/lib/python3.9/runpy.py", line 228 in run_module
  File "/Users/guilhermeleobas/git/numba/./runtests.py", line 16 in <module>
[1]    71417 segmentation fault  ./runtests.py -v -f numba.tests.test_generators.TestGenerators -k objmode
```

The order tests execute matter as running each test individually works.

guilhermeleobas commented:
CI is green

stuartarchibald commented:
/AzurePipelines run

sklam commented:
I'm testing a fix (https://github.com/numba/numba/pull/8811/commits/3d550561a2f69dc8a8f6cb2aa7de4ac25ed4a74d) for the segfault at #8811

guilhermeleobas commented:
Thanks @sklam. Patch seems to work

stuartarchibald commented on numba/core/callconv.py:
```diff
@@ -442,75 +474,86 @@ def unpack_dynamic_exception(self, builder, api, status):
+        # We check at this stage if creating the Python tuple was successful
+        # or not. Note the exception is raised by calling PyErr_SetString
+        # directly as the function being processed is the CPython wrapper
+        # to-do: fix!
+        # failed = cgutils.is_null(builder, py_tuple)
+        # with cgutils.if_unlikely(builder, failed):
+        #     msg = ('Error creating Python tuple from runtime exception '
+        #            'arguments')
+        #     pyapi.err_set_string("PyExc_RuntimeError", msg)
+        #     builder.ret(pyapi.get_null_object())
+
```
This needs addressing?

guilhermeleobas replied:
Fixed

stuartarchibald reviewed (commented):
Thanks for the updates @guilhermeleobas, there's a few largely small things left to address. The main concern I think is replacement of `ir.TryRaise` with `ir.DynamicRaise` (reasons noted inline). Thanks again!

guilhermeleobas commented:
> ok, I think the confusion here arises from the use of the type of the LLVM objects involved in the predicate guarding this block. Is the question essentially "Can this object be boxed", if not raise?

Yes! That's the purpose of the code above. I will update the comment to better explain the use case.

guilhermeleobas commented on numba/core/untyped_passes.py:
```diff
@@ -1724,3 +1724,34 @@ def _patch_locals(self, state):
+                dyn_raise = ir.DynamicRaise(exc_type, tuple(exc_args), loc)
+                block.insert_after(dyn_raise, raise_)
+                block.remove(raise_)
+        return False
```
Fixed

stuartarchibald commented on numba/core/callconv.py:
```diff
@@ -492,14 +493,12 @@ def unpack_dynamic_exception(self, builder, pyapi, status):
+        with cgutils.if_unlikely(builder, failed):
+            msg = ('Error creating Python tuple from runtime exception '
+                   'arguments')
+            pyapi.err_set_string("PyExc_RuntimeError", msg)
```
```suggestion
            pyapi.err_set_string("PyExc_RuntimeError", msg)
            builder.ret(pyapi.get_null_object())
```
I'm not sure that this has any effect unless the error state is communicated through a return value of NULL.

stuartarchibald commented on numba/core/callconv.py:
```diff
@@ -572,6 +571,8 @@ def emit_unwrap_dynamic_exception_fn(self, module, st_type, nb_types):
         st_type_ptr = st_type.as_pointer()
         st_ptr = builder.bitcast(fn.args[0], st_type_ptr)
         # compile time values are stored as None
+        # TODO: It should be possible to remove "nb_types" as argument to this
+        # function and get the Numba type from the LLVM type
```
```suggestion
```
I'm not sure this is possible as Numba has types that alias to the same LLVM type.

stuartarchibald commented on numba/core/callconv.py:
```diff
@@ -580,10 +581,10 @@ def emit_unwrap_dynamic_exception_fn(self, module, st_type, nb_types):
             if obj == cgutils.get_null_value(obj.type):
                 # When not supported, abort compilation
-                msg = f'cannot convert native {typ} to python object'
+                msg = f'Cannot convert native {typ} to a Python object.'
```
There is no unit test for this condition?

guilhermeleobas replied:
Yes, there is. See `test_exceptions::TestRaising::test_dynamic_raise_bad_args`

stuartarchibald replied:
Ah, thanks, apologies, completely missed that (I skimmed the diff of the last two commits for two updates and missed the second in the test suite).

guilhermeleobas commented on numba/core/ir.py:
```diff
@@ -816,12 +817,19 @@ def __init__(self, exc_class, exc_args, loc):
+            return f"{self.Kind}_try_raise {self.exc_class}({args})"
+
+
+class DynamicTryRaise(StaticTryRaise):
+    """A raise statement inside a try-block.
+    Similar to ``DynamicRaise`` but does not terminate.
+    """
+    Kind = 'dynamic'
```
`DynamicTryRaise` has the same methods and values as `StaticTryRaise`. If needed, I can just copy instead of inheriting.

stuartarchibald replied:
I think copy or create a common base, either is fine. Was just trying to avoid what I think would seem to be an unexpected inheritance relationship.

stuartarchibald commented on numba/core/lowering.py:
```diff
@@ -621,6 +624,11 @@ def lower_setitem(self, target_var, index_var, value_var, signature):
         return impl(self.builder, (target, index, value))
 
+    def lower_try_dynamic_raise(self, inst):
+        # Numba is a bit limited in what it can do with exceptions in a try
+        # block. Thus, it is safe to use the same code as the static try raise.
+        self.lower_static_try_raise(inst)
```
Thanks for confirming @sklam.

stuartarchibald commented on numba/tests/test_exceptions.py:
```diff
@@ -430,6 +430,22 @@ def raise_with_no_nrt(i):
         with self.assertRaisesRegex(errors.NumbaRuntimeError, msg):
             raise_with_no_nrt(123)
 
+    def test_try_raise(self):
+
+        @njit
+        def raise_(a):
+            raise ValueError(a)
+
+        @njit
+        def try_raise(a):
+            try:
+                raise_(a)
+            except Exception:
+                pass
+            return a + 1
+
+        self.assertEqual(try_raise.py_func(3), try_raise(3))
```
I've got to what I think is the cause of this. Essentially when the code for unwrapping the dynamic exception is generated it has to emit code to box native representations into python objects, but the "environment manager" is missing as kwarg "`env_manager=`" here:
https://github.com/numba/numba/blob/68b0d750eb32e2e1eed50a94062e9bb33c85b52d/numba/core/callconv.py#L582

You can get an environment manager by calling `self.context.get_env_manager(builder)` but this isn't enough though because the emitted environment sentry needs to be aware that this is a Python context code region. I made this change locally to be able to wire this through:
```diff
@@ -123,13 +123,14 @@ class CPUContext(BaseContext):
             builder, envptr, _dynfunc._impl_info['offsetof_env_body'])
         return EnvBody(self, builder, ref=body_ptr, cast_ref=True)
 
-    def get_env_manager(self, builder):
+    def get_env_manager(self, builder, return_pyobject=False):
         envgv = self.declare_env_global(builder.module,
                                         self.get_env_name(self.fndesc))
         envarg = builder.load(envgv)
         pyapi = self.get_python_api(builder)
         pyapi.emit_environment_sentry(
             envarg, debug_msg=self.fndesc.env_name,
+            return_pyobject=return_pyobject
         )
         env_body = self.get_env_body(builder, envarg)
         return pyapi.get_env_manager(self.environment, env_body, envarg)

```
and then did this:
```diff
@@ -576,10 +576,12 @@ class CPUCallConv(BaseCallConv):
         nb_types = [typ for typ in nb_types if typ is not None]
 
         # convert native values into CPython objects
+        env_manager = self.context.get_env_manager(builder,
+                                                   return_pyobject=True)
         objs = []
         for i, typ in enumerate(nb_types):
             val = builder.extract_value(builder.load(st_ptr), i)
-            obj = pyapi.from_native_value(typ, val)
+            obj = pyapi.from_native_value(typ, val, env_manager=env_manager)

```
thoughts?

sklam replied:
oh, that's fine

guilhermeleobas replied:
Thanks @stuartarchibald

stuartarchibald replied:
> Thanks @stuartarchibald

No problem.

guilhermeleobas commented on numba/core/ir.py:
```diff
@@ -776,7 +805,28 @@ class StaticTryRaise(Stmt):
+            return f"static_try_raise {self.exc_class}({args})"
+
 
+class DynamicTryRaise(StaticTryRaise):
```
Oops, my mistake. Fixed!

guilhermeleobas commented on numba/core/callconv.py:
```diff
@@ -413,6 +466,260 @@ def return_user_exc(self, builder, exc, exc_args=None, loc=None,
+        for i, typ in enumerate(nb_types):
+            val = builder.extract_value(builder.load(st_ptr), i)
+            env_manager = self.context.get_env_manager(builder,
+                                                       return_pyobject=True)
```
Fixed

stuartarchibald commented:
I've checked/reviewed up to b1eb24b, looks good. Given this enables "dynamic" exceptions, last task is to remove the restriction note from the docs:

https://github.com/numba/numba/blob/2829f1f108d3350f456984036f9742c95ce41bfc/docs/source/reference/pysupported.rst?plain=1#L142-L143
(suggest just making it read `raise SomeException(<arguments>)`, i.e. no restriction clause)

stuartarchibald reviewed (dismissed):
@guilhermeleobas It's great to see this feature make it into Numba, thanks for all your work on it and for your persistence in getting it completed!

Patch approved pending buildfarm testing!

guilhermeleobas commented:
Thanks for all the help. It was a good journey to work on this feature.

stuartarchibald reviewed (approved):
Thanks for the fixes @guilhermeleobas.

stuartarchibald commented:
Buildfarm ID: `numba_smoketest_cpu_yaml_170`.

gmarkall commented:
gpuci run tests

gmarkall reviewed (approved):
I haven't reviewed the code, but this has my approval based on the gpuCI pass (i.e. it's not breaking CUDA exceptions).

sklam commented:
`numba_smoketest_cpu_yaml_170` passed

dlee992 commented:
Hi, just leave a comment: this PR definitely helps us for debugging. Thanks! However, in some cases, it does hurt the execution time when with this PR. In principle, you think in which way this PR can affect the performance, compared with the old constant way?

guilhermeleobas commented:
Hi @dlee992, glad this work is helping you.

> In principle, you think in which way this PR can affect the performance, compared with the old constant way?

Cannot think of an example of why this can cause a performance bottleneck. It may increase compilation time, but it shouldn't affect the execution. Do you have a reproducer that I can take a look at?

dlee992 commented:
Ah, I retried again. When applying this PR locally, I also brought in the changes in #8535 in `callconv.py`. After I removed the changes from 8535 (sth related to changing the `_return_errcode_raw`), the execution time seems back to normal.

So accidentally, maybe I found the root cause of my another issue #9186. Maybe it's not llvm14 or this PR's fault, changes in `_return_errcode_raw` could be.
